### PR TITLE
DBZ-6240 Change stream supports filtering schema change event

### PR DESCRIPTION
**Pull request by sunxiaojian**
Change stream supports filtering schema change event

### CODE REVIEW

**Comment by jpechane** [comment]
@sunxiaojian Hi, thanks for the PR. I am not fully convinced we want to add such functionality directly to the Debezium internals. Is this something that could be achieve via an SMT? That would be used when needed and be configured to filter out requested DDLs. At the same time it would be applicble at any connector.

**Comment by sunxiaojian** [comment]
> @sunxiaojian Hi, thanks for the PR. I am not fully convinced we want to add such functionality directly to the Debezium internals. Is this something that could be achieve via an SMT? That would be used when needed and be configured to filter out requested DDLs. At the same time it would be applicble at any connector.

It's ok add an SMT to filter, but I think it should be used with the include.schema.changes configuration

**Comment by sunxiaojian** [comment]
> We tend to avoid too many config options if possible. Generally only if a performance gain can be obtained which is not the case of DDL filtering.

Okay, I'll add an SMT to filter

**Review by Naros — COMMENTED** [review]
Overall this LGTM @sunxiaojian, there is just one outstanding task: adding a reference to this new transformation in the documentation.

**Comment by sunxiaojian** [comment]
> Overall this LGTM @sunxiaojian, there is just one outstanding task: adding a reference to this new transformation in the documentation.

Okay, let me add it

File: debezium-core/src/main/java/io/debezium/transforms/SchemaChangeEventFilter.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+public class SchemaChangeEventFilter<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SchemaChangeEventFilter.class);
+
+    private static final Field SCHEMA_CHANGE_EVENT_INCLUDE_LIST = Field.create("schema.change.event.include.list")
+            .withDisplayName("Schema change event include list")
+            .withType(ConfigDef.Type.LIST)
+            .withWidth(ConfigDef.Width.SHORT)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDescription(
+                    "Support filtering during DDL synchronization!");
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/transforms/SchemaChangeEventFilter.java:43]:
```suggestion
                    "Support filtering during DDL synchronization");
```

File: debezium-core/src/main/java/io/debezium/transforms/SchemaChangeEventFilter.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    private Set<SchemaChangeEvent.SchemaChangeEventType> includeSchemaChangeEvents;
+    private SmtManager<R> smtManager;
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        final Configuration config = Configuration.from(configs);
+        smtManager = new SmtManager<>(config);
+
+        final String includeSchemaChangeEvents = config.getString(SCHEMA_CHANGE_EVENT_INCLUDE_LIST);
+        if (includeSchemaChangeEvents == null) {
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/transforms/SchemaChangeEventFilter.java:54]: Please take a look at `io.debezium.transforms.HeaderToValue.HEADERS_FIELD` for better way how to handle field validation.

File: debezium-core/src/main/java/io/debezium/transforms/SchemaChangeEventFilter.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            // throw exception
+            throw new IllegalArgumentException("The configuration [schema.change.event.include.list] cannot be empty");
+        }
+        this.includeSchemaChangeEvents = (includeSchemaChangeEvents == null)
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/transforms/SchemaChangeEventFilter.java:58]: IIUC there lines above ensure that `includeSchemaChangeEvents` will be alway non-null.

**Review by jpechane — CHANGES_REQUESTED** [review]
@sunxiaojian Thanks for that. I left two more minor comments to address.

**Comment by sunxiaojian** [comment]
@jpechane @Naros PTAL

**Review by jpechane — CHANGES_REQUESTED** [review]
@sunxiaojian Overall, the code LGTM!
I did few changes in docs. Alos I've one request for you - the name `schema.change.event.include.list` is very confusing as it implies the list of DDL types that should be kept in the stream but IIUC they are in fact removed. So I propose renaming it to either `schema.change.event.exclude.list` or `schema.change.event.filter.list`.

@roldanbob Could you please review the docs as well? I am unsure about the formatting a bit.

**Comment by sunxiaojian** [comment]
> @sunxiaojian Overall, the code LGTM! I did few changes in docs. Alos I've one request for you - the name `schema.change.event.include.list` is very confusing as it implies the list of DDL types that should be kept in the stream but IIUC they are in fact removed. So I propose renaming it to either `schema.change.event.exclude.list` or `schema.change.event.filter.list`.
> 
> @roldanbob Could you please review the docs as well? I am unsure about the formatting a bit.

@jpechane   Okay , I have modified it to 'schema.change.event.exclude.list'

**Comment by roldanbob** [comment]
> @roldanbob Could you please review the docs as well? I am unsure about the formatting a bit.

@jpechane Should I approach this as if it's going to be exposed downstream (that is, add the necessary Nebel annotations)?

**Review by roldanbob — CHANGES_REQUESTED** [review]
@sunxiaojian  Thank you for providing this new documentation. I've suggested some language, style, and formatting changes. 
I tested the formatting changes in a local version of the file, but to guard against the possibility that I missed anything, I'd like to review the changes again after you make updates. 
I've also posted a couple of questions related to fields in the properties table.

Let me know if you have any concerns or questions about any edits.

**Comment by sunxiaojian** [comment]
> @sunxiaojian Thank you for providing this new documentation. I've suggested some language, style, and formatting changes. I tested the formatting changes in a local version of the file, but to guard against the possibility that I missed anything, I'd like to review the changes again after you make updates. I've also posted a couple of questions related to fields in the properties table.
> 
> Let me know if you have any concerns or questions about any edits.

@roldanbob done

File: documentation/modules/ROOT/pages/transformations/schema-change-event-filter.adoc [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+== Effect of applying the `SchemaChangeEventFilter` SMT
+After you configure the SMT for a connector, whenever the connector captures `DROP` or `TRUNCATE` events, it resets the values for these events to `null`.
+The connector does not send events with the resulting `null` values to Kafka.
+The following examples show how the transformation changes the values of event records that you designate for exclusion.
+.Schema change event record before it is processed by the `SchemaChangeEventFilter` transformation
+====
+Schema change event record before it is processed by the `SchemaChangeEventFilter` transformation::
+
+Schema change event record before the SMT processes the event record:::
++
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/schema-change-event-filter.adoc:44]:
```suggestion
The connector does not send events with the resulting `null` values to Kafka.

The following examples shows a xref:schema-change-event-filter-example-before-processing[schema change event record before it is processed] by the SMT, followed by the xref:schema-change-event-filter-example-after-processing[output after the SMT filters the event].

[[schema-change-event-filter-example-before-processing]]
.Schema change event record before it is processed by the `SchemaChangeEventFilter` transformation
====
```

File: documentation/modules/ROOT/pages/transformations/schema-change-event-filter.adoc [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+.Schema change event record after processing by the SMT
+====
+
+[source]
+----
+null
+----
+====
+
+
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/schema-change-event-filter.adoc:122]:
```suggestion

// Type: reference
```

File: documentation/modules/ROOT/pages/transformations/schema-change-event-filter.adoc [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                "type":"DROP",
+                "id":"test.table"
+            }
+        ]
+    }
+}
+----
+
+====
+.Schema change event record after processing by the SMT
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/schema-change-event-filter.adoc:113]:
```suggestion

Based on the earlier xref:example-configuration-schema-change-event-filter[configuration example], after the SMT processes the xref:schema-change-event-filter-example-before-processing[original schema change event], it resets the values for the `DROP` event to `null`, as shown in the following example:

[[schema-change-event-filter-example-after-processing]]
.Schema change event record after processing by the SMT
```

**Review by roldanbob — CHANGES_REQUESTED** [review]
Added some internal cross-references, removed obsolete formatting, and inserted downstream metadata comments.

**Comment by jpechane** [comment]
@sunxiaojian Applied, thanks!

**Pull request merged**